src/legacy/core_plugins/kibana/public/visualize/plugin.ts

```diff
@@ -70,46 +75,72 @@ export class VisualizePlugin implements Plugin {
     share: SharePluginStart;
     visualizations: VisualizationsStart;
   } | null = null;
+  private appStateUpdater = new BehaviorSubject<AngularRenderedAppUpdater>(() => ({}));
+  private stopUrlTracking: (() => void) | undefined = undefined;
 
   constructor(private initializerContext: PluginInitializerContext) {}
 
   public async setup(
     core: CoreSetup,
-    { home, kibanaLegacy, __LEGACY, usageCollection }: VisualizePluginSetupDependencies
+    { home, kibanaLegacy, usageCollection, data }: VisualizePluginSetupDependencies
   ) {
+    const { querySyncStateContainer, stop: stopQuerySyncStateContainer } = getQueryStateContainer(
+      data.query
+    );
+    const { appMounted, appUnMounted, stop: stopUrlTracker, setActiveUrl } = createKbnUrlTracker({
+      baseUrl: core.http.basePath.prepend('/app/kibana'),
+      defaultSubUrl: '#/visualize',
+      storageKey: 'lastUrl:visualize',
+      navLinkUpdater$: this.appStateUpdater,
+      toastNotifications: core.notifications.toasts,
+      stateParams: [
+        {
+          kbnUrlKey: '_g',
+          stateUpdate$: querySyncStateContainer.state$,
+        },
+      ],
+    });
+    this.stopUrlTracking = () => {
+      stopQuerySyncStateContainer();
+      stopUrlTracker();
+    };
+
     kibanaLegacy.registerLegacyApp({
       id: 'visualize',
       title: 'Visualize',
+      updater$: this.appStateUpdater.asObservable(),
+      navLinkId: 'kibana:visualize',
       mount: async (params: AppMountParameters) => {
         const [coreStart] = await core.getStartServices();
```

Comment:
@lizozom  @flash1293  
If I'm not wrong we decided to use approach with creating  a **services.ts** file and pass start dependencies using **createGetterSetter** method.

Reply:
I wasn't aware this is a definitive rule, just an option how to do things - passing down services directly can IMHO be preferable, depending on the situation (e.g. because you don't have to mock out the getters in tests)

Reply:
Looks like this comment is not related to this PR but here we pass some dependencies using `core.getStartServices()`; but some comes from this.startDependencies object.

Is it possible to refactor it in future and do it in a common with other plugins way?